### PR TITLE
rpc: Add RPC help for getblock verbosity level 3

### PR DESCRIPTION
**Pull request by kiminuo**
This is a follow-up PR to #22918 which addresses review comments (first commit). The second commit adds missing RPC help for verbosity level 3.

### CODE REVIEW

**Comment by theStack** [comment]
Concept ACK

**Comment by laanwj** [comment]
Any specific reason for closing?

**Comment by kiminuo** [comment]
@laanwj No, that was just a mistake as I was deleting old branches and this one was deleted by mistake too. Sorry.

**Comment by laanwj** [comment]
@kiminuo No problem! Will take a look it seems very close to ready to merge.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -984,6 +985,37 @@ static RPCHelpMan getblock()
...
+                    {RPCResult::Type::ARR, "tx", "",
+                    {
+                        {RPCResult::Type::OBJ, "", "",
+                        {
+                            {RPCResult::Type::ARR, "vin", "",
+                            {
+                                {RPCResult::Type::OBJ, "", "",
+                                {
+                                    {RPCResult::Type::ELISION, "", "The same output as verbosity = 2"},
+                                    {RPCResult::Type::OBJ, "prevout", "",
```
Comment (laanwj) [inline src/rpc/blockchain.cpp:1001]: Might want to mention briefly this is only available if there is undo information for the input (e.g., not for coinbase inputs). E.g. a user might be surprised to not see it when running it on the genesis block.

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -984,6 +985,37 @@ static RPCHelpMan getblock()
...
+                                    {
+                                        {RPCResult::Type::BOOL, "generated", "Coinbase or not"},
+                                        {RPCResult::Type::NUM, "height", "The height of the prevout"},
+                                        {RPCResult::Type::NUM, "value", "The value in " + CURRENCY_UNIT},
+                                        {RPCResult::Type::OBJ, "scriptPubKey", "",
+                                        {
+                                            {RPCResult::Type::STR, "asm", "The asm"},
+                                            {RPCResult::Type::STR, "hex", "The hex"},
+                                            {RPCResult::Type::STR, "address", /* optional */ true, "The Bitcoin address (only if a well-defined address exists)"},
+                                            {RPCResult::Type::STR, "type", "The type, eg 'pubkeyhash'"},
```
[the remark this answers is not in the document]
Comment (laanwj) [reply src/rpc/blockchain.cpp:1011]: SGTM, it cannot hurt to enumerate them if a function to provide that information is available already anyway.

**Comment by laanwj** [comment]
Code review and lightly tested ACK 08d129a406947b39f7822458ba9a1af30aeacbc3
Small nit on the documentation but feel free to ignore.

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (kiminuo) [reply src/rpc/blockchain.cpp:1001]: Happy to change. It's just about finding the proper sentence. What about 

```cpp
{RPCResult::Type::OBJ, "prevout", "(Only if undo information is available)",
```

?

Or is it too brief?

**Comment by pg156** [comment]
`src/bitcoin-cli help getblock` doesn't return anything related to "verbosity = 3" on 
macOS Big Sur. Did I do something wrong in the steps below?

```
git clone https://github.com/kiminuo/bitcoin.git 23320
cd 23320
git checkout feature/2021-10-20-verbosity-level-3-nits      
./autogen.sh
./configure --with-incompatible-bdb --without-miniupnpc --without-natpmp --disable-bench --disable-wallet --without-gui
le-bdb --without-miniupnpc --without-natpmp --disable-bench --disable-wallet --without-gui
make -j 3 src/bitcoin-cli
./src/bitcoin-cli help getblock

**Comment by theStack** [comment]
> `src/bitcoin-cli help getblock` doesn't return anything related to "verbosity = 3" on macOS Big Sur. Did I do something wrong in the steps below?

@pg156: You have to build and run `bitcoind`, as the help texts to the RPC calls are residing in there. `bitcoin-cli` is merely the RPC client that fetches it, i.e. most likely have another instance running on your system that has nothing to do with this PR. For simple tests like those, I recommend running bitcoind with `-nolisten -noconnect` to prevent unnecessary network activity.

**Comment by pg156** [comment]
Thank you @theStack! I get "verbosity = 3" after building `bitcoind` and stopping another `bitcoind` running on my machine, as you suggested.

**Review by pg156 — COMMENTED** [review]
Does this line also need to be updated for verbosity = 3?

https://github.com/bitcoin/bitcoin/blob/1bdd5f63229ebf28c24a8656f486ed8a6c8b3787/src/rpc/blockchain.cpp#L946

**Comment by kiminuo** [comment]
@pg156 It's certainly better to update it. Thank you. Any ideas about better wordings are welcome :)

**Review by CryptoGGytrs52 — APPROVED** [review]

**Comment by pg156** [comment]
Code review ACK https://github.com/bitcoin/bitcoin/commit/3a6a670456b0f45edc8c70647fd49969adef2e74.
(but https://github.com/bitcoin/bitcoin/commit/3a6a670456b0f45edc8c70647fd49969adef2e74 doesn't exist in the "Commits" tab after the force-push. I don't know if the standard practice is to ACK only what is in the tab?)

- Tested the difference in output of `getblock` between verbosity of 2 and 3 is as described in RPC help, by adding two lines of logging in `_test_getblock` function in `test/functional/rpc_blockchain.py`:

```python
self.log.info(node.getblock(blockhash, 2))
self.log.info(node.getblock(blockhash, 3))
```

The difference is:

```python
'prevout': {'generated': True, 'height': 18, 'value': Decimal('50.00000000'), 'scriptPubKey': {'asm': '0 4ae81572f06e1b88fd5ced7a1a000945432e83e1551e6f721ee9c00b8cc33260', 'hex': '00204ae81572f06e1b88fd5ced7a1a000945432e83e1551e6f721ee9c00b8cc33260', 'address': 'bcrt1qft5p2uhsdcdc3l2ua4ap5qqfg4pjaqlp250x7us7a8qqhrxrxfsqseac85', 'type': 'witness_v0_scripthash'}}
```

- Tested "verbosity = 3" section in RPC help is as expected by running:

```shell
./src/bitcoind -testnet -daemon         
./src/bitcoin-cli -testnet help getblock
```

which generates:

```shell
...
Result (for verbosity = 3):
{                                        (json object)
  ...,                                   Same output as verbosity = 2
  "tx" : [                               (json array)
    {                                    (json object)
      "vin" : [                          (json array)
        {                                (json object)
          ...,                           The same output as verbosity = 2
          "prevout" : {                  (json object) (Only if undo information is available)
            "generated" : true|false,    (boolean) Coinbase or not
            "height" : n,                (numeric) The height of the prevout
            "value" : n,                 (numeric) The value in BTC
            "scriptPubKey" : {           (json object)
              "asm" : "str",             (string) The asm
              "hex" : "str",             (string) The hex
              "address" : "str",         (string, optional) The Bitcoin address (only if a well-defined address exists)
              "type" : "str"             (string) The type, eg 'pubkeyhash'
            }
          }
        },
        ...
      ]
    },
    ...
  ]
}
```

**Comment by kiminuo** [comment]
> (but [3a6a670](https://github.com/bitcoin/bitcoin/commit/3a6a670456b0f45edc8c70647fd49969adef2e74) doesn't exist in the "Commits" tab after the force-push. I don't know if the standard practice is to ACK only what is in the tab?)

Typically one ACKs the last commit from https://github.com/bitcoin/bitcoin/pull/23320/commits page because that's the current version of the PR.

You can also possibly ACK a commit from previous git push to say something like "I really like the previous changeset as opposed to the current changeset" but that would be rare AFAIK.

**Comment by kiminuo** [comment]
@pg156 Could you please re-review?
@MarcoFalke Would you mind having a look, please?

**Comment by pg156** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/23320/commits/059f88b6a97b7a3ae1f033885e40ac01f91e6d60

File: src/rpc/blockchain.cpp (hunk 1 above)
Comment (laanwj) [reply src/rpc/blockchain.cpp]: Seems fine to me, agree that the set of conditions is hard to describe precisely in the scope available.

**Comment by laanwj** [comment]
re-ACK 059f88b6a97b7a3ae1f033885e40ac01f91e6d60

**Pull request merged**